File: CS/Exercises/Exercise5/Runner.cs
using System.Xml;
using System.Xml.Linq;

namespace Exercises.Exercise5;

class Runner
{
  private const string _connectionString = @"Server=.\profitsqldev;Database=demo;Trusted_Connection=true;TrustServerCertificate=True";
  private const string _xmlFileName = @"c:\temp\demo.xml";

  public static void RunMe()
  {
    var persons = new AntaDataBuffer<PersonRow>(row => row.Id.Value);

    var elements = StreamXml(_xmlFileName);

    foreach (var personElement in elements.Take(100))
    {
      var personRow = new PersonRow();
      personRow.Id.Value = new Guid(personElement.Attribute("id")!.Value);
      personRow.FirstName.Value = personElement.Element("FirstName")!.Value;
      personRow.LastName.Value = personElement.Element("LastName")!.Value;
      persons.Add(personRow);

      WriteToSql(personRow);
    }

    Console.WriteLine("Loaded {0} persons.", persons.Count);
    Console.ReadKey();
  }

  private static void WriteToSql(PersonRow personRow)
  {
  }

  private static IEnumerable<XElement> StreamXml(string uri)
  {
    using (var reader = XmlReader.Create(uri))
    {
      reader.ReadStartElement("Persons");

      while (reader.Read())
      {
        if (reader.NodeType == XmlNodeType.Element)
        {
          yield return (XElement.ReadFrom(reader) as XElement)!;
        }
      }

      reader.Close();
    }
  }
}

File: VB/Exercises/Exercise5/Runner.vb
Imports System.Xml

Namespace Exercises.Exercise5
  Class Runner
    Private Const _connectionString As String = "Server=.\profitsqldev;Database=demo;Trusted_Connection=true;TrustServerCertificate=True"
    Private Const _xmlFileName = "c:\temp\demo.xml"

    Public Shared Sub RunMe()
      Dim persons = New AntaDataBuffer(Of PersonRow)(Function(row) row.Id.Value)

      Dim elements = StreamXml(_xmlFileName)

      For Each personElement In elements.Take(100)
        Dim personRow = New PersonRow()
        personRow.Id.Value = New Guid(personElement.Attribute("id").Value)
        personRow.FirstName.Value = personElement.Element("FirstName").Value
        personRow.LastName.Value = personElement.Element("LastName").Value
        persons.Add(personRow)

        WriteToSql(personRow)
      Next

      Console.WriteLine("Loaded {0} persons.", persons.Count)
      Console.ReadKey()

    End Sub


    Private Shared Sub WriteToSql(personRow As PersonRow)

    End Sub

    Private Shared Iterator Function StreamXml(uri As String) As IEnumerable(Of XElement)
      Using reader As XmlReader = XmlReader.Create(uri)
        reader.ReadStartElement("Persons")

        While reader.Read()
          If (reader.NodeType = XmlNodeType.Element) Then
            Yield CType(XElement.ReadFrom(reader), XElement)
          End If
        End While

        reader.Close()
      End Using
    End Function
  End Class
End Namespace
